packages/plugins/typescript/gql-tag-operations/src/index.ts

```diff
@@ -21,13 +21,34 @@ export type DocumentType<TDocumentNode extends DocumentNode<any, any>> = TDocume
   : never;
 `;
 
+const maskingTypePartial = `
+export type FragmentType<TDocumentType extends DocumentNode<any, any>> = TDocumentType extends DocumentNode<
+  infer TType,
+  any
+>
+  ? TType extends { ' $fragmentName': infer TKey }
+    ? TKey extends string
+      ? { ' $fragmentRefs': { [key in TKey]: TType } }
+      : never
+    : never
+  : never;
+
+export const useFragment = <TType>(
+  _documentNode: DocumentNode<TType, any>,
+  fragmentType: FragmentType<DocumentNode<TType, any>>
+): TType => {
+  return fragmentType as any;
+};
```

Comment:
If frameworks wanna implement true fragment type masking at runtime they just need to implement this function type signature.